src/pool-cl/lens/CLQuoter.sol

```diff
@@ -5,288 +5,162 @@ import {TickMath} from "pancake-v4-core/src/pool-cl/libraries/TickMath.sol";
 import {ICLPoolManager} from "pancake-v4-core/src/pool-cl/interfaces/ICLPoolManager.sol";
 import {BalanceDelta} from "pancake-v4-core/src/types/BalanceDelta.sol";
 import {PoolKey} from "pancake-v4-core/src/types/PoolKey.sol";
-import {PoolIdLibrary} from "pancake-v4-core/src/types/PoolId.sol";
+import {PoolId, PoolIdLibrary} from "pancake-v4-core/src/types/PoolId.sol";
 import {ICLQuoter} from "../interfaces/ICLQuoter.sol";
 import {PoolTicksCounter} from "../libraries/PoolTicksCounter.sol";
 import {PathKey, PathKeyLibrary} from "../../libraries/PathKey.sol";
-import {Quoter} from "../../base/Quoter.sol";
+import {BaseV4Quoter} from "../../base/BaseV4Quoter.sol";
+import {QuoterRevert} from "../../libraries/QuoterRevert.sol";
+import {Currency} from "pancake-v4-core/src/types/Currency.sol";
 
-contract CLQuoter is Quoter, ICLQuoter {
+contract CLQuoter is ICLQuoter, BaseV4Quoter {
+    using QuoterRevert for *;
     using PathKeyLibrary for PathKey;
+    using PoolIdLibrary for PoolId;
 
     ICLPoolManager public immutable poolManager;
-    uint256 private constant CL_MINIMUM_VALID_RESPONSE_LENGTH = 192;
 
-    /// @dev min valid reason is 6-words long (192 bytes)
-    /// @dev int128[2] includes 32 bytes for offset, 32 bytes for length, and 32 bytes for each element
-    /// @dev Plus sqrtPriceX96After padded to 32 bytes and initializedTicksLoaded padded to 32 bytes
-    constructor(address _poolManager) Quoter(_poolManager, CL_MINIMUM_VALID_RESPONSE_LENGTH) {
+    constructor(address _poolManager) BaseV4Quoter(_poolManager) {
         poolManager = ICLPoolManager(_poolManager);
     }
 
     /// @inheritdoc ICLQuoter
     function quoteExactInputSingle(QuoteExactSingleParams memory params)
         external
-        override
-        returns (int128[] memory deltaAmounts, uint160 sqrtPriceX96After, uint32 initializedTicksLoaded)
+        returns (uint256 amountOut, uint256 gasEstimate)
     {
+        uint256 gasBefore = gasleft();
         try vault.lock(abi.encodeCall(this._quoteExactInputSingle, (params))) {}
         catch (bytes memory reason) {
-            return _handleRevertSingle(reason);
+            gasEstimate = gasBefore - gasleft();
```

Comment:
The gas estimation is accurate than v2/stableSwap because of try-revert manner but still not 100% accurate because in real trade there would be  some periphery logic & token settement